Fixes Massive Radio Overtime, Implements a Spatial Grid System for Faster Searching Over Areas

Kylerace opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
a month or two ago i realized that on master the reason why get_hearers_in_view() overtimes so much (ie one of our highest overtiming procs at highpop) is because when you transmit a radio signal over the common channel, it can take ~20 MILLISECONDS, which isnt good when 1. player verbs and commands usually execute after SendMaps processes for that tick, meaning they can execute AFTER the tick was supposed to start if master is overloaded and theres a lot of maptick 2. each of our server ticks are only 50 ms, so i started on optimizing this. 

the main optimization was SSspatial_grid which allows searching through 17x17 spatial_grid_cell datums (one set for each z level) far faster than iterating over movables in view() to look for what you want. now all hearing sensitive movables in the 5x5 areas associated with each spatial_grid_cell datum are stored in the datum (so are client mobs). when you search for one of the stored "types" (hearable or client mob) in a radius around a center, it just needs to 
1. iterate over the cell datums in range
2. add the content type you want from the datums to a list
3. subtract contents that arent in range, then contents not in line of sight
4. return the list

from benchmarks, this makes short range searches like what is used with radio code (it goes over every radio connected to a radio channel that can hear the signal then calls get_hearers_in_view() to search in the radios canhear_range which is at most 3) about 3-10 times faster depending on workload. the line of sight algorithm scales well with range but not very well if it has to check LOS to > 100 objects, which seems incredibly rare for this workload, the largest range any radio in the game searches through is only 3 tiles

the second optimization is to enforce complex setter vars for radios that removes them from the global radio list if they couldnt actually receive any radio transmissions from a given frequency in the first place. 

the third optimization i did was massively reduce the number of hearables on the station by making hologram projectors not hear if dont have an active call/anything that would make them need hearing. so one of hte most common non player hearables that require view iteration to find is crossed out. 

also implements a variation of an idea oranges had on how to speed up get_hearers_in_view() now that ive realized that view() cant be replicated by a raycasting algorithm. it distributes pregenerated abstract /mob/oranges_ear instances to all hearables in range such that theres at max one per turf and then iterates through only those mobs to take advantage of type-specific view() optimizations and just adds up the references in each one to create the list of hearing atoms, then puts the oranges_ear mobs back into nullspace. this is about 2x as fast as the get_hearers_in_view() on master

Fixes #62962
Fixes #62835
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
holy FUCK its fast. like really fucking fast. the only costly part of the radio transmission pipeline i dont touch is mob/living/Hear() which takes ~100 microseconds on live but searching through every radio in the world with get_hearers_in_radio_ranges() -> get_hearers_in_view() is much faster, as well as the filtering radios step

[benchmark (2).txt](https://github.com/tgstation/tgstation/files/7149178/benchmark.2.txt)
heres the benchmarking code i did here, at least part of it i forgot to save the timing on individual radio transmissions but it started at about 20ms like on live (most of the cost isnt sending to players, its just searching through random radios that likely wont even reach players)

![Screenshot_1278](https://user-images.githubusercontent.com/15794172/132977459-e07e4e11-c4ec-4662-8548-b8f14776db5e.png)

the spatial grid searching proc is about 36 microseconds/call at 10 range and 16 microseconds at 3 range in the captains office (relatively many hearables in view), the new get_hearers_in_view() was 4.16 times faster than get_hearers_in_view_old() at 10 range and 4.59 times faster at 3 range

![Screenshot_1279](https://user-images.githubusercontent.com/15794172/132977462-b5f246bb-00a8-453f-97b2-6a66a787a184.png)

in this spot, a more open area with less hearables in view (middle of the tram tracks on tramstation), the new get_hearers_in_view() was 10x faster than the old one at 10 range and 4x faster at 3 range

in both cases going over all ~250 radios in the global radio list (before i cut down on the ones that shouldnt be there it was around 600) the new get_hearers_in_radio_ranges() was 3 times faster than the old one

![Screenshot_1280](https://user-images.githubusercontent.com/15794172/132977922-d0d39be1-9bf7-4413-9309-f813b14869b3.png)
also heres some stats i made for a single radio transmission with the optimized code. about 4-5x better than it used to be

SSspatial_grid could be used for a lot more things other than just radio and say code, i just didnt implement it. for example since the cells are datums you could get all cells in a radius then register for new objects entering them then activate when a player enters your radius. this is something that would require either very expensive view() calls or iterating over every player in the global list and calling get_dist() on them which isnt that expensive but is still worse than it needs to be

on normal get_hearers_in_view cost the new version that uses /mob/oranges_ear instances is about 2x faster than the old version, especially since the number of hearing sensitive movables has been brought down dramatically. 
![Screenshot_1589](https://user-images.githubusercontent.com/15794172/143726123-687b6a71-0b2b-49bf-b13e-140c84f8672d.png)
with get_hearers_in_view_oranges_ear() being the benchmark proc that implements this system and get_hearers_in_view() being a slightly optimized version of the version we have on master, get_hearers_in_view_as() being a more optimized version of the one we have on master, and get_hearers_in_LOS() being the raycasting version currently only used for radios because it cant replicate view()'s behavior perfectly.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kylerace committed:
setters and docs for radios, removes from global lists if not listening

Kylerace committed:
stash commit to push to origin

Kylerace committed:
Revert "stash commit to push to origin"

This reverts commit 7ac4ef6b348066c6b49bd8c0bab5baf2812c6b66.

Kylerace committed:
makes client_mobs_in_contents instead use important_recursive_contents

Kylerace committed:
adds SSspatial_hashmap and associated datums and procs

Kylerace committed:
the hashmap searching proc is now 10x faster

Kylerace committed:
get_hearers_in_view() uses the hashmap and is 10x faster

Kylerace committed:
as anything in playsound

Kylerace committed:
moves another thing to update the client map correctly

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
makes the cells datums, optimizes it significantly: 15 microseconds/call

Kylerace committed:
renames it from a hashmap to a gridmap/grid since it doesnt use hashes

Kylerace committed:
gives get_hearers_in_view() an include_source argument

Kylerace committed:
autodoc comment for stack_trace() because i saw it

Kylerace committed:
cuts out 90% benchmarking code from spatial_gridmap.dm

Kylerace committed:
enforces setter and getter use for radio vars, cuts down

 on radios in the global telecomms list, which reasonably reduces the number of radios that are queried when transmitting over common

Kylerace committed:
as anything in OnTransitZ() because i saw it

Kylerace committed:
implements turf/Entered and Exited to add movables to the spatial grid

Kylerace committed:
better naming in send_speech() and makes it include src in get_hearers

Kylerace committed:
cleans up and optimizes radio code more

Kylerace committed:
makes the spatial grid not hang references

Kylerace committed:
docs for get_hearers_in_view() and renames get_mobs_in_radio_ranges0

Kylerace committed:
very helpful comment in wires code

Kylerace committed:
includes spatial_gridmap.dm

LemonInTheDark commented:
I like this. A lot. Linda needed cpu time
> 51x51 spatial_grid_cell datums

51? you mean per z level?
> the line of sight algorithm scales well with range but not very well if it has to check LOS to > 100 objects, which seems incredibly rare for this workload, the largest range any radio in the game searches through is only 3 tiles

How bad does it get? would it be worth it to set a "threshold" at which we just use view()?

Do you think it would be worth it to make radios into a subsystem to reduce the random spurts of usage?

Kylerace commented:
> > 51x51 spatial_grid_cell datums
> 
> 51? you mean per z level?

51*51 per z level. i intend to make it "lazy" in some way since the vast majority of even the station ones are empty but i havent thought of a good solution that doesnt lower performance an uncomfortable amount. at the very least i could make the datum cells contents lists lazy

> > the line of sight algorithm scales well with range but not very well if it has to check LOS to > 100 objects, which seems incredibly rare for this workload, the largest range any radio in the game searches through is only 3 tiles
> 
> How bad does it get? would it be worth it to set a "threshold" at which we just use view()?

i thought of that but decided against it for now since it only gets bad over like 100 objects to sort through and the current workload simply should not do that except in massive edge cases. perhaps in the future. it shouldnt ever be slower than how it used to be except in even more unlikely edge cases (consider a ring of open turfs enclosed by closed turfs with radius 8 and a few hundred objects distributed just outside of the closed turfs, searching through the center with range 10 should be slower than the old get_hearers_in_view() because view() wouldnt return any of them in the first place but my algorithm has to check every one)

> Do you think it would be worth it to make radios into a subsystem to reduce the random spurts of usage?

could be, but for now ill settle for making it cheap as fuck (although 5 ms isnt really cheap in absolute terms, so yeah id say it could definitely use yielding before the expensive steps)

> ~Please tell me this isn't stolen from goon I hate it when that happens~

100% of this code is my own

LemonInTheDark reviewed (changes requested):
I'm kinda curious how you intended on optimizing this, and what the like, typical usecase ought to look like

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+
+/datum/controller/subsystem/spatial_grid/Initialize(start_timeofday)
+	. = ..()
+	var/cells_per_side = world.maxx / SPATIAL_GRID_CELLSIZE //assume world.maxx == world.maxy
```
then you change SPATIAL_GRID_CELLSIZE to be dividable. and if someone makes world.maxx a prime number we let them be stupid

LemonInTheDark replied:
Why not just make this division round up? so if it's like 51.3 it just goes to 54?

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+	///what z level we are in
+	var/cell_z
+
+	//every data point in a grid cell is separated by usecase
+
+	///every hearing sensitive movable inside this cell
+	var/list/hearing_contents = list()
+	///every client possessed mob inside this cell
+	var/list/client_contents = list()
```
define generically

LemonInTheDark replied:
Rather then a new list per type of contents, a list of lists indexed by type.
I guess what's important is how easily it to be for people to add new categories. Reading through the comments it sounds like the answer to that is reasonably hard

LemonInTheDark replied:
Also this should maybe be a lazylist

Kylerace replied:
i dont think it should be made into a single list, making associative indexes is very expensive and movement is already a massive drain on performance, most of that being player movement. a player moving from a cell containing only themselves to an empty cell is making and deleting one big list, and two associative indexes and two sub lists as opposed to adding to two lists currently. not to mention it will slow reads.

if someone wants to add a new content type they should be able to do so without compromising performance in the slightest, especially if more systems start to rely on the grid. its essentially just adding their new content define to the gridmap defines folder then adding it to the switch statements (as well as the list correlated with that define). i should probably make the switch statements a define though

Kylerace replied:
also the content lists are now lazy

LemonInTheDark replied:
pogged up

LemonInTheDark commented:
> 51*51 cells per z level. i intend to make it "lazy" in some way since the vast majority of even the station ones are empty but i havent thought of a good solution that doesnt lower performance an uncomfortable amount. at the very least i could make the datum cells contents lists lazy

With 9 z levels that ends up being like, 23409 cells doesn't it? or am I doing my math wrong

> i thought of that but decided against it for now since it only gets bad over like 100 objects to sort through and the current workload simply should not do that except in massive edge cases. perhaps in the future. it shouldnt ever be slower than how it used to be except in even more unlikely edge cases (consider a ring of open turfs enclosed by closed turfs with radius 8 and a few hundred objects distributed just outside of the closed turfs, searching through the center with range 10 should be slower than the old get_hearers_in_view() because view() wouldnt return any of them in the first place but my algorithm has to check every one)

Aight, we ought to be able to play it roughly by ear
 
> could be, but for now ill settle for making it cheap as fuck (although 5 ms isnt really cheap in absolute terms, so yeah id say it could definitely use yielding before the expensive steps)
I love you
 
> 100% of this code is my own

see above, but combine it with mind seethe directed at zewaka

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+	var/min_y = max(CEILING((center_y - range) * (1 / SPATIAL_GRID_CELLSIZE), 1), 1)//calculating these indices only takes around 2 microseconds
+
+	//the maximum x and y cell indexes to test
+	var/max_x = min(CEILING((center_x + range) * (1 / SPATIAL_GRID_CELLSIZE), 1), grid_cells_per_axis)
+	var/max_y = min(CEILING((center_y + range) * (1 / SPATIAL_GRID_CELLSIZE), 1), grid_cells_per_axis)
```
as deep as we can go, backing out a little bit if we dig too deep and too greedily

LemonInTheDark commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+				for(var/datum/spatial_grid_cell/cell as anything in grid_row.Copy(min_x, max_x + 1))
+					contents_to_return += cell.hearing_contents
+
+	if(!include_center) //this is faster for things that dont care about themselves but are (probably) in the output, helps us return without using the LOS algorithm
+		contents_to_return -= center
```
Are we expecting to do a lot of -=ling/how large is the contents to return list?

Kylerace replied:
usually less than 10 elements as far as im aware. this check is so that you dont have to go through the line of sight algorithm if you dont get anything except yourself and you dont want to do anything with yourself (like radios)

LemonInTheDark replied:
Aight. if we expected a large amount of items I'd tell you to make a new list of filtered things to return for the range/view loops, so we don't need to iterate over the list each time to remove an entry.

I wonder if iterating by index would be faster. that's too in the weeds though

Kylerace replied:
i thought that 
```
for(var/list/row as anything in grid.Copy(min_y, max_y + 1))
	for(var/datum/spatial_grid_cell/cell as anything in row.Copy(min_y, max_y + 1))
```
was faster from some old benchmarks, but in those i didnt realize that the last arg of Copy is one off from what i thought, so it was actually going through less cells. now it iterates through index which makes the proc ~10% faster. it also has to check if the list exists now which makes it ~4% slower (not an actual measurement), so it should be a little bit faster now and work with lazy content lists.

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+		//maybe in the future a high volume algorithm would be worth it
+		var/turf/inbetween_turf = center_turf
+		while(TRUE)
+			inbetween_turf = get_step(inbetween_turf, get_dir(inbetween_turf, target_turf))
+
+			if(inbetween_turf == target_turf)//we've gotten to target's turf without returning due to turf opacity, so we must be able to see target
+				break
+
+			if(inbetween_turf.opacity || inbetween_turf.opacity_sources)//this turf or something on it is opaque so we cant see through it
+				contents_to_return -= target
+				break
```
get_step(ref, diagonal dir) works

LemonInTheDark replied:
yeah, but it'll also put you through a wall.
So
```
W = wall
V = viewer
T = target
VW
WT
```
Would incorrectly say it's fine, right? God this problem sucks.

Ghilker replied:
this should be a mapping issue since usually mappers don't make these kind of structures

Ghommie replied:
But players can ingame so it's not simply a mapping issue.

LemonInTheDark replied:
Enforcing that on mappers wouldn't be justified. This game is a sandbox, if we need to lay down restrictions like that for our code to function our code is bad.

Also we'd need to lint maps in a really annoying way, which I don't want to do

Kylerace replied:
Does view() prefer cardinals over diagonals like our movement code?

LemonInTheDark replied:
view works the same way as the black "dead zones" do doesn't it?

LemonInTheDark replied:
outside of returning items outside in

Kylerace replied:
im definitely not introducing a view() call here. i might be able to think of something but right now im not sure how to make it completely accurate to view() without being view() or being slow.

LemonInTheDark replied:
It's this problem right?
![image](https://user-images.githubusercontent.com/58055496/136124763-02cf6c61-1e2d-4f28-98cc-5aecdd57f9d4.png)
I think I've ran into it recently. Solution looked like this, see the comments on /datum/move_loop/has_target/move_towards
https://github.com/LemonInTheDark/tgstation/blob/meteor-subsystem/code/controllers/subsystem/movement/movement_types.dm#L318

Now that I look at it it still has some flaws. We're not always guaranteed to move 1 tile at a time, there should be logic to prevent x = 1, y = 0.5 sorts of things. Could also almost for sure be done better with trig functions, more a prototype then anything else.

This might also be helpful, if a bit incomplete. There's lots of people trying to solve this same issue, one of their solutions will work.

If you figure it out replace https://github.com/LemonInTheDark/tgstation/blob/meteor-subsystem/code/__HELPERS/unsorted.dm#L431. Man's a bastard

Kylerace replied:
ok, now get_hearers_in_view() works with view() again, and get_hearers_in_LOS() has been made to interface with the spatial grid. anything that isnt important to be 1:1 with view() uses the LOS version

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+
+	var/list/contents_to_return = list()
+
+	var/static/grid_cells_per_axis = world.maxx / SPATIAL_GRID_CELLSIZE//im going to assume this doesnt change at runtime
```
> 
> With 9 z levels that ends up being like, 23409 cells doesn't it? or am I doing my math wrong
> 
yep, as well as 9 + (9 * 51) + (23409 * 2) = 47286 lists

Kylerace commented:
> 
> With 9 z levels that ends up being like, 23409 cells doesn't it? or am I doing my math wrong
> 
yep, as well as 9 + (9 * 51) + (23409 * 2) = 47286 lists

optimumtact commented:
can you explain the spatial grid more, is it a tree structure or just a bigger grid.

What cost is paid to update things positions in the spatial grid

optimumtact commented:
Another location that would benefit is the code that wakes up simple ai's when things enter their ranges or the z level.

Kylerace commented:
> 
> 
> can you explain the spatial grid more, is it a tree structure or just a bigger grid.

theres a list of z level grids, each one of those is a list of 51 "row" lists ie every y coordinate is a list of 51 spatial_grid_cell datums which are what actually contains the contents

> What cost is paid to update things positions in the spatial grid

turf/Entered() adds to the cells and turf/Exited() subtracts from the cell if 1. the last turf you were in was covered by a different grid cell and 2. you can be put into the grid cell in the first place.
im not sure exactly how expensive this is, i tested it briefly and it seemed cheap enough but since movement is currently one of the biggest costs on the server i should check what percentage of Move's cost it is

optimumtact commented:
> > can you explain the spatial grid more, is it a tree structure or just a bigger grid.
> 
> theres a list of z level grids, each one of those is a list of 51 "row" lists ie every y coordinate is a list of 51 spatial_grid_cell datums which are what actually contains the contents

So there's only a row grid? there's no x component? Or there's a list of x per z level with y's inside each x entry?

> 
> > What cost is paid to update things positions in the spatial grid
> 
> turf/Entered() adds to the cells and turf/Exited() subtracts from the cell if 1. the last turf you were in was covered by a different grid cell and 2. you can be put into the grid cell in the first place.

can you precalculate which turfs need to care about this (or is it not worth it since the math should be a pretty simple `y%n == 0 or 1` = im a border turf

Kylerace commented:
@optimumtact 

> 
> So there's only a row grid? there's no x component? Or there's a list of x per z level with y's inside each x entry?
> 

there is an x component inside each row list, the cells are stored in order of x index inside each row list. x and y indexes are just x and y coordinates divided by the cell size (5) and rounded up. for example, a turf at coordinates (121, 130, 3) is in the 3rd z level grid map in the grids_by_z_level list, 130/5 = 26th row list inside that z level grid, and the 121/5 rounded up = 25th spatial_grid_cell datum inside that row list. eg the spatial cell of any turf is 
```
grids_by_z_level[z][CEILING(y / 5, 1)][CEILING(x / 5, 1)]
```

> 
> can you precalculate which turfs need to care about this (or is it not worth it since the math should be a pretty simple `y%n == 0 or 1` = im a border turf

i could put vars on each turf of what their index in the grid is, id have to put it on every turf though since for non adjacent movement you can always skip straight into the middle of a different cell instead of the border. currently i calculate this in turf/Entered() and turf/Exited(), but yes the operation to calculate whether the old turf is in the same grid cell as the new one should be pretty cheap. heres the one in turf/Entered() 
```
var/turf/old_turf = get_turf(old_loc)

if(old_turf?.z == z && CEILING(old_turf.x / SPATIAL_GRID_CELLSIZE, 1) == CEILING(x / SPATIAL_GRID_CELLSIZE, 1) && CEILING(old_turf.y / SPATIAL_GRID_CELLSIZE, 1) == CEILING(y / SPATIAL_GRID_CELLSIZE, 1))
	return
```
i was curious how much memory two index vars on every turf would take for 9 z levels and the answer is that it would actually take more than everything else im currently doing unless im doing my math wrong. using 
![image](https://user-images.githubusercontent.com/15794172/133191771-15310b8f-124d-45c5-9a2e-1535d29ead6f.png)
255 world.maxx * 255 world.maxy * 9 z levels * 2 changed vars * 12 bytes per changed var = 14 megabytes

while the entire spatial grid should take 23409 cell datums * 24 bytes per datum + 5 changed vars per datum * 12 bytes per changed var * 23409 datums + 47287 total lists * 24 bytes per list + 9 z level grid elements * 51 row lists per z level grid * 51 datum cells per row list * 8 bytes per list element = 3.2 megabytes

optimumtact commented:
yeah you're right, and I don't think there's a spare turf flag slot, so it's fine

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+		if(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS)
+			for(var/list/grid_row as anything in grid_level.Copy(min_y, max_y + 1))//from testing, slightly faster than iterating numbers from min_index to max_index
+				for(var/datum/spatial_grid_cell/cell as anything in grid_row.Copy(min_x, max_x + 1))
+					contents_to_return += cell.client_contents
```
🤔

Kylerace replied:
im pretty sure it is since vars is also a var, so youre doing a GetVar then pushing the name then a ListGet instead of just a GetVar. at least according to crucible
![Screenshot_1377](https://user-images.githubusercontent.com/15794172/136124388-634bd39b-4946-459b-b986-963591f78bbe.png)
![Screenshot_1376](https://user-images.githubusercontent.com/15794172/136124392-c9dff06f-8756-4a87-b38c-6f4146c494ee.png)

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+	var/static/grid_cells_per_axis = world.maxx / SPATIAL_GRID_CELLSIZE//im going to assume this doesnt change at runtime
+
+	//the minimum x and y cell indexes to test
+	var/min_x = max(CEILING((center_x - range) * (1 / SPATIAL_GRID_CELLSIZE), 1), 1)
```
are you sure its not already evaluated once?

LemonInTheDark replied:
no

Kylerace replied:
done

Kylerace committed:
gets rid of TODOs that i forgot about

Kylerace committed:
makes the datum content lists lazy, optimizes the searching slightly

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+	cell.client_contents -= to_remove
+	cell.hearing_contents -= to_remove
+
+///if shit goes south, this will find hanging references for qdeleting movables inside
```
forgot to make it actually remove from the cells, now an admin could theoretically do that however i dont have a way to detect whether this is a problem automatically.

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+
+	var/list/intersecting_grid_cells = list()
+
+	var/static/grid_cells_per_axis = world.maxx / SPATIAL_GRID_CELLSIZE//im going to assume this doesnt change at runtime
```
now should be consistent

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+
+	var/list/grid = grids_by_z_level[target_turf.z]
+
+	var/datum/spatial_grid_cell/cell_to_return = grid[CEILING(target_turf.y / SPATIAL_GRID_CELLSIZE, 1)][CEILING(target_turf.x / SPATIAL_GRID_CELLSIZE, 1)]
```
now should be consistent

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,256 @@
+	var/list/grid_level = grids_by_z_level[center_turf.z]
+	switch(type)
+		if(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS)
+			for(var/list/grid_row as anything in grid_level.Copy(min_y, max_y + 1))//from testing, slightly faster than iterating numbers from min_index to max_index
```
now it goes by index again, no use of Copy() (its reasonably faster)

Kylerace committed:
fixes dumb runtimes

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
did i seriously leave this code non compilable for a week

LemonInTheDark commented:
You can't take this from me

Kylerace commented:
working on wallening (and school). will go back to this soonish

Kylerace committed:
re adds cell_z spatial_grid_cell var for metadata

Kylerace committed:
HOPEFULLY fixes client contents and entering the grid pre init

Kylerace committed:
makes the spatial grid cell size 15 instead of 5 for better performance

Kylerace committed:
adds ROUND_UP() define to replace CEILING(value, 1)

Kylerace committed:
makes the gridmap cells NON LAZY again to speed up searching

Kylerace committed:
major changes to grid searching, also the grid signals are sent to the subsystem instead of the cells

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
makes SSspatial_grid register to new z levels creating. fix unit tests?

Kylerace committed:
hopefully fixes fucking test failures. remakes client_mobs_in_contents and replaces recursive contents trait signal handlers

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
LEMON CUCKED ME AGAIN REEEEE FIXES IT TEMPORARILY

Kylerace committed:
optimizes and makes get_line() more readable because i was looking there

Kylerace committed:
makes SSspatial_grid check tick in Initialize()

Kylerace committed:
gets rid of dumb INVERSE_SPATIAL_GRID_CELLSIZE and makes stuff work with the normal one because i figured out that dividing by x is just as fast as multiplying by 1/x in byond :trollface:

Kylerace committed:
makes the spatial cell entering and exiting signals less stupid

Kylerace committed:
fixes infinite loop, temporarily fixes poly for TM

Kylerace committed:
fixes bug with removing things from the grid before its initialized

Kylerace committed:
gets rid of a braindead restriction on speaking into headsets that was losing input due to the new changes

listening doesnt correlate with not being able to speak into the headset and have it transmit what you said, it just means the headset isnt relaying radio transmissions to you. now poly can speak on radio

Kylerace committed:
adds vars to track whether radios "should" be listening/broadcasting etc

also rearranges the radio procs to be in a more sane order

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
adds client safety change from master that caused merge conflict

Kylerace committed:
actually fixes the pre init removal runtime by returning

Kylerace committed:
:trollface:

Kylerace committed:
adds doc comment explaining decision

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
separates get_hearers into raycasting and view variants and optimizes

Kylerace committed:
fixes bug with updating the spatial grid in Moved()

Kylerace committed:
comment stuff

Kylerace committed:
gets rid of dumb old stupid commented out code

Kylerace committed:
fixes headsets not working in non human mobs

Kylerace committed:
makes telecomms code slightly less horrifically stupid

creates a new list to allow better filtering of links to only the relevant types

Kylerace committed:
gets rid of debug message

Kylerace committed:
adds support for expanding the grid if world.maxx or maxy increases

Kylerace committed:
makes vocal/broadcast less susceptible to nulls in lists since its vital

Kylerace committed:
makes the cell content lists point to a single empty list if empty

this is meant to save memory because there arent two empty instantiated lists for every cell, now theyre just references to a single list in SSspatial_grid and become separate lists if something is added to them and become references again if everything is removed.

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
fixes merge problems and syndicate radio bug

Kylerace committed:
makes holopads only be hearing sensitive when needed

holopads are a very common hearing sensitive movable on station, which leads to fetching hearables take longer because of more view iterating in the case of get_hearers_in_view() and more raycasting steps in the case of get_hearers_in_LOS(). lowering the density of hearables on station also makes tree-based datastructures like a quadtree more viable to implement which might be capable of replacing the grid.

Kylerace committed:
refactors get_hearers_in_view() to use a new system thats ~2x faster

it now distributes pregenerated abstract /mob/oranges_ear instances to all hearable atoms in range and gives them references to all hearables on their tile, then iterates through just the oranges_ear instances in view() which saves on list creation cost and leads to much faster view speed. its still a third as fast as get_hearers_in_LOS() which doesnt use the grid lol

Kylerace committed:
removes unneeded defines for holopads

Kylerace committed:
nullchecks send_speech and adds temporary procs to profile certain costs

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
makes send_speech skip over nulls in the hearing movables list

Kylerace committed:
makes /mob/oranges_ear return an initialization hint

Kylerace committed:
fixes oranges ears

Kylerace committed:
gets rid of useless lists for oranges_ear mobs

LemonInTheDark commented on code/__HELPERS/game.dm:
```diff
@@ -238,10 +238,48 @@
 	var/list/references = list()
 
-/mob/orange_ear/Initialize(mapload)
+/mob/oranges_ear/Initialize(mapload)
```
Lol

Kylerace replied:
fake news

Kylerace committed:
hopefully fixes tests

Kylerace committed:
adds more functionality to find_grid_statistics_for_z_level

Kylerace committed:
makes the grid cells size 17 since number of contents has decreased

i didnt actually benchmark how much of a difference it makes but with 100 players assigned to random station turfs only like a third of all cells have contents

Kylerace committed:
improves sanic speeds

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace commented on code/__HELPERS/maths.dm:
```diff
@@ -30,34 +30,44 @@
+	var/y_distance_sign = SIGN(y_distance)
+
+	var/x = abs_x_distance >> 1 //Counters for steps taken, setting to distance/2
+	var/y = abs_y_distance >> 1 //Bit-shifting makes me l33t.  It also makes get_line() unnessecarrily fast.
```
(it didnt make it faster according to tests but i didnt want to steal his thunder)

Kylerace replied:
(also two divisions isnt going to be a major component of this procs cost anyways)

LemonInTheDark replied:
lol

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
gets rid of [] syntax in get_line() and debug procs

Kylerace committed:
adds more autodocs for the spatial grid and pals

LemonInTheDark commented on code/__HELPERS/game.dm:
```diff
@@ -213,26 +213,189 @@
+	///used for guaranteeing there is only one oranges_ear per turf when assigned, speeds up view() iteration
+	var/mob/oranges_ear/assigned_oranges_ear
+
+//turns out view() spends a significant portion of its processing time generating lists of contents of viewable turfs which includes EVERYTHING on it visible
+//and the turf itself. there is an optimization to view() which makes it only generate lists of a certain atom type - this system takes advantage of that.
+//a fuckton of these are generated as part of its SS's init and stored in a list, when requested for a list of movables returned by the spatial grid or by some
+//superset of the final output that must be narrowed down by view() one of these gets put on every turf that contains the movables that need filtering
+//and each is given references to the movables they represent. that way you can do for(var/mob/oranges_ear/ear in view(...)) and check what they reference
+//as opposed to for(var/atom/movable/target in view(...)) and checking if they have the properties you want which leads to much larger lists generated by view()
+//and also leads to iterating through more movables to filter them.
+//TLDR: iterating through just mobs is much faster than all movables when iterating through view(), this system leverages that to boost speed
+
```
Autodock block comment?

Kylerace replied:
done

LemonInTheDark commented on code/__HELPERS/game.dm:
```diff
@@ -213,26 +213,189 @@
+	/// movables iside of other movables count as being "on" if they have get_turf(them) == our turf. intentionally not a lazylist
+	var/list/references = list()
+
+/mob/oranges_ear/Initialize(mapload)
+	SHOULD_CALL_PARENT(FALSE)
+	return INITIALIZE_HINT_NORMAL
```
Why no destroy?

Kylerace replied:
added

LemonInTheDark commented on code/__HELPERS/game.dm:
```diff
@@ -213,26 +213,189 @@
+
+///clean this oranges_ear up for future use
+/mob/oranges_ear/proc/unassign()
+	loc:assigned_oranges_ear = null//trollface. our loc should ALWAYS be a turf, no exceptions. if it isnt then this doubles as an error message ;)
```
If you do this I can't bully goon over having a : in their flow code
Is this really so hot you can't afford a typecast, even with no check

LemonInTheDark commented on code/__HELPERS/game.dm:
```diff
@@ -213,26 +213,189 @@
+	//on a whole this can outperform iterating through all movables in view() by ~2x especially when hearables are a tiny percentage of movables in view
+	for(var/mob/oranges_ear/ear in view(view_radius, center_turf))
+		. += ear.references
+		ear.unassign()
+		assigned_oranges_ears -= ear
+
+	for(var/mob/oranges_ear/remaining_ear as anything in assigned_oranges_ears)//we need to clean up our mess
+		remaining_ear.unassign()
```
Why not just do all the cleaning on one for loop?

Kylerace replied:
thats a good catch it should be faster to do it in the second loop. especially since it doesnt require subtracting from it

Kylerace replied:
done

LemonInTheDark commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,567 @@
+
+// macros meant specifically to add/remove movables from the hearing_contents and client_contents lists of
+// /datum/spatial_grid_cell, when empty they become references to a single list in SSspatial_grid and when filled they become their own list
+// this is to save memory without making them lazylists as that slows down iteration through them
```
assertion here is insertion/modification is less of a hot path then reading, right?

Kylerace replied:
most likely. i dont actually think this system is needed at all anymore since all our calculations for how much memory this system was taking were made when the cells were 5x5 but now theyre 17x17. meaning instead of 2601 cells per z level theres only 225

Kylerace replied:
no wait i misread this, most of the time since players move a lot im pretty sure contents move between cells more often than cells are read

LemonInTheDark commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,567 @@
+	client_contents = dummy_list
+
+/datum/spatial_grid_cell/Destroy(force, ...)
+	if(force)//the response to someone trying to qdel this is a right proper fuck you
```
stack trace, I wish to be able to yell at people if they do it in ci

Kylerace replied:
done

Kylerace committed:
brain crashing rapidly optimizes oranges_ear stuff again

Kylerace committed:
separates spatial helpers from game.dm into spatial_info.dm

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into nonshit-telecomms

Kylerace committed:
fixes zombie error

Kylerace committed:
adds doc comment and Destroy() to /mob/oranges_ear

Kylerace committed:
inlines oranges_ear/assign() since its used in one place and is slow

Kylerace committed:
adds a stack_trace() to spatial_grid_cell/Destroy()

Kylerace committed:
asserts dominance over the spatial grid system file

LemonInTheDark commented:
Ready?

Kylerace commented:
@LemonInTheDark ready

LemonInTheDark commented:
Set go

LemonInTheDark commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -0,0 +1,593 @@
+			turfs = GLOB.station_turfs
+
+		else
+			turfs = block(locate(1,1,z), locate(255,255,z))
```
oh one more thing, this shouldn't be hardcoded, right?

Kylerace replied:
probably not, fixed

Kylerace committed:
makes the statistics proc use world.maxx and maxy

Cyberboss commented:
based

optimumtact commented:
@Cyberboss based on what??

Kylerace commented:
based on the spatial data structure its embedded in

LemonInTheDark commented:
Based on orange_ear
